packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -2299,6 +2327,68 @@ export class StripeHelper {
     return this.stripe[resourceType].retrieve(resource);
   }
 
+  async processWebhookEventToFirestore(event: Stripe.Event) {
+    const { type, data } = event;
+
+    if (!this.stripeFirestore) {
+      return;
+    }
+
+    // Note that we must insert before any event handled by the general
+    // webhook code to ensure the object is up to date in Firestore before
+    // our code handles the event.
+    try {
+      switch (type as Stripe.WebhookEndpointUpdateParams.EnabledEvent) {
+        case 'invoice.created':
+        case 'invoice.finalized':
+        case 'invoice.paid':
+        case 'invoice.payment_failed':
+        case 'invoice.updated':
+        case 'invoice.deleted':
+          const invoice = data.object as Stripe.Invoice;
+          await this.stripeFirestore.retrieveAndFetchSubscription(
+            invoice.subscription as string
+          );
+          // Now that the subscription is fetched, we can safely insert the
+          // invoice record.
+          await this.stripeFirestore.insertInvoiceRecord(invoice);
+          break;
+        case 'customer.created':
+        case 'customer.updated':
+        case 'customer.deleted':
+          const customer = data.object as Stripe.Customer;
+          // Ensure the customer and its subscriptions exist in Firestore.
+          // Note that we still insert the object here in case we've already
+          // fetched the customer previously.
+          await this.stripeFirestore.retrieveAndFetchCustomer(customer.id);
+          await this.stripeFirestore.insertCustomerRecord(
+            customer.metadata.userid,
+            customer
+          );
```

Comment:
This is to handle `customer.updated` and `customer.deleted`, in which case we already have a customer record (that is now stale) that would be returned by `retrieveAndFetchCustomer` without updating the customer in Firestore?

Reply:
Given that we are not immediately back-filling Firestore, its quite likely we can get deleted events for people we have never inserted. This will cause many of these events to fail as they can't retrieve sufficient customer information if Stripe has already deleted it all. That's why this entire block is wrapped in a try/except to catch that scenario.

Reply:
Okay thanks; sorry my comment did not indicate the right set of lines here, as I was asking about calling `retrieveAndFetchCustomer` before calling `insertCustomerRecord`.